### PR TITLE
[WIP] Enforce TransferRate Maximum

### PR DESCRIPTION
**Pull request by xasos**
Requires and amendment change. Sets a maximum value for TransferRate at 2 billion, rather than the current assumed value (~4.29 billion). Treats any requested TransferRate over the limit as the max value. cc @nbougalis @scottschurr

### CODE REVIEW

File: src/ripple/ledger/impl/View.cpp [hunk 1]
```diff
@@ -391,7 +391,11 @@ transferRate (ReadView const& view,
     auto const sle = view.read(keylet::account(issuer));
 
     if (sle && sle->isFieldPresent (sfTransferRate))
+    {
+        if (sfTransferRate > 2 * QUALITY_ONE)
```
Comment (nbougalis) [inline src/ripple/ledger/impl/View.cpp]: This doesn't do what you expect it to. You're not retrieving the transfer rate entry.

File: src/ripple/ledger/impl/View.cpp [hunk 2]
```diff
@@ -391,7 +391,11 @@ transferRate (ReadView const& view,
...
+            return Rate{ 2 * QUALITY_ONE };
         return Rate{ sle->getFieldU32 (sfTransferRate) };
+    }
```
Comment (nbougalis) [inline src/ripple/ledger/impl/View.cpp]: So, you could collapse all this to something like:

```return Rate{ std::min((*sle)[sfTransferRate], 2 * QUALITY_ONE) };```

But since we want this code to be on an amendment, we'll need something more verbose:

Consider:
```auto const rate = (*sle)[sfTransferRate];
if (/*amendmentActive*/ && rate > maxRate)
    rate = std::min(rate, maxRate);
return Rate{rate};
```

Comment (xasos) [inline src/ripple/ledger/impl/View.cpp]: View.rules().enabled(fix1201)

File: src/test/rpc/AccountSet_test.cpp [hunk 3]
```diff
@@ -211,13 +211,19 @@ class AccountSet_test : public beast::unit_test::suite
 
         uint32 xfer_rate = 2000000000;
         jt[sfTransferRate.fieldName] = xfer_rate;
-        env(jt);
+        env(jt, ter(temBAD_TRANSFER_RATE));
```
Comment (nbougalis) [inline src/test/rpc/AccountSet_test.cpp]: If we're setting the transfer rate to the maximum allowed, why would this fail? It seems like it should not.

Also, we should test two cases: amendment disabled (so, the _old_ behavior should be allowed) and amendment enabled (so the _new_ behavior is allowed).

Comment (nbougalis) [inline src/test/rpc/AccountSet_test.cpp]: If we're setting the transfer rate to the maximum allowed, why would this fail? It seems like it should not.

Also, we should test two cases: amendment disabled (so, the _old_ behavior should be allowed) and amendment enabled (so the _new_ behavior is allowed).

File: src/test/app/TransferRate_test.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        env(pay("alice", "bob", USD(1)));
+        env.require(balance("alice", USD(aliceBal)));
+        env.require(balance("bob", USD(1)));
+    }
+
+    void
+    run() override
+    {
+        test(1000000000, 1);
+        test(2000000000, 0);
```
Comment (nbougalis) [inline src/test/app/TransferRate_test.cpp]: The tests here need to be a bit more extensive: we need to test with the amendment disabled (so, the _old_ behavior is allowed) and with the amendment enabled.

Additionally, the function should take the full set of parameters (that is, how much to send from Alice to Bob, and how much Alice and Bob should have in case of success). Right now, we mix things up a bit - some of the values are in hard-coded in the function and others in the invocation.

File: src/ripple/ledger/impl/View.cpp [hunk 5]
```diff
@@ -391,7 +391,13 @@ transferRate (ReadView const& view,
     auto const sle = view.read(keylet::account(issuer));
 
     if (sle && sle->isFieldPresent (sfTransferRate))
-        return Rate{ sle->getFieldU32 (sfTransferRate) };
+    {
+        auto rate = (*sle)[sfTransferRate];
+        auto maxRate = 2 * QUALITY_ONE;
```
Comment (nbougalis) [inline src/ripple/ledger/impl/View.cpp]: You can skip the `maxRate` here and go for something like this:

    rate = std::min(rate, 2 * QUALITY_ONE);

**Comment by scottschurr** [comment]
I'll look at this harder on Monday, but first a quick observation.  I noted that the pull request is hanging off of `ripple:master`.  Seems like a better place would be `ripple:develop`?

File: src/ripple/ledger/impl/View.cpp (hunk 5 above)
Comment (xasos) [inline src/ripple/ledger/impl/View.cpp]: std::min requires both numbers to be the same type - since rate is a uint32_t, and 2 * QUALITY_ONE is a signed int, we have to make them uniform. I ended up removing the maxRate instantiation and casting it as an unsigned int.

However, I'd vote for keeping in maxRate because it makes the code a bit more readable.

**Comment by xasos** [comment]
@scottschurr Once this commit is ready to go, I'll squash+cherry pick the commits and open a PR to develop

File: src/test/app/TransferRate_test.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+// Make sure transfer rate 
+class TransferRate_test : public beast::unit_test::suite
+{
+public:
+    void
+    test (std::uint32_t tr, std::uint32_t aliceBal)
+    {
+        using namespace jtx;
+        Env env(*this);
```
Comment (mellery451) [inline src/test/app/TransferRate_test.cpp]: very minor: we've been tending to prefer brace construction in newer code.

File: src/test/app/TransferRate_test.cpp (hunk 4 above)
Comment (mellery451) [inline src/test/app/TransferRate_test.cpp]: I agree with Nik about covering the boundary conditions better, and I would also try to write them mostly in terms of the QUALITY_ONE constant.

File: src/test/app/TransferRate_test.cpp (hunk 6 above)
Comment (xasos) [inline src/test/app/TransferRate_test.cpp]: For consistency's sake, going to leave it with parens, but will keep in mind for the future.

File: src/ripple/ledger/impl/View.cpp (hunk 5 above)
Comment (nbougalis) [inline src/ripple/ledger/impl/View.cpp]: I am convinced re: readability. The signedness is an issue that should be addressed in another commit in a better way.

File: src/test/rpc/AccountSet_test.cpp [hunk 7]
```diff
@@ -211,13 +211,19 @@ class AccountSet_test : public beast::unit_test::suite
 
         uint32 xfer_rate = 2000000000;
```
Comment (scottschurr) [inline src/test/rpc/AccountSet_test.cpp]: Consider changing `2000000000` to `QUALITY_ONE*2`.

File: src/test/app/TransferRate_test.cpp (hunk 4 above)
Comment (scottschurr) [inline src/test/app/TransferRate_test.cpp]: I agree with the preceding comments.  I'd be curious to see a bunch of test cases.  You might start with:
- 0
- 1
- `QUALITY_ONE` - 1
- `QUALITY_ONE`
- `QUALITY_ONE`*2
- `QUALITY_ONE`*2 + 1
- `std::numeric_limits<std::uint32_t>::max()`.

Then try those four cases with and without the amendment enabled.

Then try applying the transfer rate to `alice` (as you're currently doing), but also to `gw` alone and also to both `alice` and `gw`.  I suspect applying it to `gw` is more typical of what we'd expect.  But it's also good to see what happens if a non-issuer (like alice) sets a transfer rate.

FWIW, there is a (tiny) pre-existing transfer rate test in AccountSet_test.cpp.  Might it make sense for your test(s) to go in there?  I'm not sure that putting your tests in a new file is a net win...

I just looked a bit more at AccountSet_test.cpp.  A few of the tests I've suggested are covered there.  That may argue for keeping all of the transfer rate tests in the same file so it's a bit easier to see when there's redundancy.

File: src/test/app/TransferRate_test.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        auto const gw = Account("gateway");
+        auto const USD = gw["USD"];
+        env.fund(XRP(10000), gw, "alice", "bob");
+        env.trust(USD(2), "alice", "bob");
+        env.close();
+		auto jt = noop("alice");
```
Comment (xasos) [inline src/test/app/TransferRate_test.cpp]: Hmm, not sure why that's there. Probably some weird whitespace issue with my editor.

File: src/ripple/ledger/impl/View.cpp [hunk 9]
```diff
@@ -391,7 +391,13 @@ transferRate (ReadView const& view,
     auto const sle = view.read(keylet::account(issuer));
 
     if (sle && sle->isFieldPresent (sfTransferRate))
-        return Rate{ sle->getFieldU32 (sfTransferRate) };
+    {
+        auto rate = (*sle)[sfTransferRate];
+        uint32_t maxRate = 2 * QUALITY_ONE;
+        if (view.rules().enabled(fix1201))
```
Comment (scottschurr) [inline src/ripple/ledger/impl/View.cpp]: Just a quick note that you'll need to add a `#include <ripple/protocol/Feature.h>` to the top of View.cpp for a non-unity build to succeed.

File: src/ripple/ledger/impl/View.cpp [hunk 10]
```diff
@@ -391,7 +391,13 @@ transferRate (ReadView const& view,
...
+            rate = std::min(rate, maxRate);
```
Comment (scottschurr) [inline src/ripple/ledger/impl/View.cpp]: Tiny nit: consider reformulating these lines as:
```
        auto rate = (*sle)[sfTransferRate];
        if (view.rules().enabled(fix1201))
        {
            constexpr uint32_t maxRate = 2 * QUALITY_ONE;
            rate = std::min(rate, maxRate);
        }
        return Rate{rate};
```
The `constexpr` assures you that the calculation happens at compile time.  Moving `maxRate` inside the braces limits its scope to the minimum necessary.

File: src/ripple/protocol/impl/TER.cpp [hunk 11]
```diff
@@ -112,7 +112,7 @@ bool transResultInfo (TER code, std::string& token, std::string& text)
         { temBAD_SIGNATURE,          { "temBAD_SIGNATURE",         "Malformed: Bad signature."                                                     } },
         { temBAD_SIGNER,             { "temBAD_SIGNER",            "Malformed: No signer may duplicate account or other signers."                  } },
         { temBAD_SRC_ACCOUNT,        { "temBAD_SRC_ACCOUNT",       "Malformed: Bad source account."                                                } },
-        { temBAD_TRANSFER_RATE,      { "temBAD_TRANSFER_RATE",     "Malformed: Transfer rate must be >= 1.0"                                       } },
+        { temBAD_TRANSFER_RATE,      { "temBAD_TRANSFER_RATE",     "Malformed: Transfer rate must be > 1 and <= 2.0"                                       } },
```
Comment (scottschurr) [inline src/ripple/protocol/impl/TER.cpp]: Should this be `>=1.0`?

File: src/test/rpc/AccountSet_test.cpp [hunk 12]
```diff
@@ -300,6 +402,7 @@ class AccountSet_test : public beast::unit_test::suite
         testSetAndResetAccountTxnID();
         testSetNoFreeze();
         testDomain();
+        testGateway();
```
Comment (scottschurr) [inline src/test/rpc/AccountSet_test.cpp]: Two of your gateway tests are failing.  That's cool, I very much want the increased test coverage, and I know you're working on it.  But I also don't want to annoy you by pointing out things I'm sure you already know.  And I'd also like to conserve my code review eyes for when you are ready for me to look at stuff.

So could you specifically let us know when you're ready for more code review?  That way you can push to your heart's content until you are ready.  Thanks.

Comment (xasos) [inline src/test/rpc/AccountSet_test.cpp]: Of course @scottschurr, my apologies! Didn't realize you got pinged every time I pushed a new commit. Will update you + other reviewers when I'm all finished up.

**Comment by xasos** [comment]
Moved to #2148.
